rust/transaction-filter/src/filters/event_filter.rs

```diff
@@ -0,0 +1,49 @@
+use crate::{filters::MoveStructTagFilter, json_search::JsonSearchTerm, traits::Filterable};
+use anyhow::Error;
+use aptos_protos::transaction::v1::{move_type::Content, Event};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+#[serde(deny_unknown_fields)]
+pub struct EventFilter {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub data: Option<JsonSearchTerm>,
+    // Only for events that have a struct as their generic
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub struct_type: Option<MoveStructTagFilter>,
+    // TODO: address? etc
+}
+
+impl Filterable<Event> for EventFilter {
+    fn is_valid(&self) -> Result<(), Error> {
+        if self.data.is_none() && self.struct_type.is_none() {
+            return Err(Error::msg(
```

Comment:
Not 100% sure what the errors would look like but I'd suggest making sure it's obvious to the user which part of their query is invalid. If all they see is `At least one of data or struct_type must be set` it doesn't make it clear which part is the problem.

Reply:
it would be some kind of catch + rewrap... doable, can do it in a follow up PR